In [1]:
import numpy as np
import pandas as pd

In [2]:
import pandas as pd

col_Names=["Sample code number","Clump thickness", "Uniformity of cell Size", "Uniformity of cell shape", "marginal Ahesion","Single epithelial cell size", "Bare Nuclei", "Bland Chromatin", "Normal Nucleoli", "Mitoses", "Class"]

bc = pd.read_csv("C:/Users/Laghavi/Desktop/Sem1/Data Mining and Management/Assignment 5/breast-cancer-wisconsin.csv", names=col_Names)


In [3]:
bc

Sample code number  Clump thickness  Uniformity of cell Size  \
0               1000025                5                        1   
1               1002945                5                        4   
2               1015425                3                        1   
3               1016277                6                        8   
4               1017023                4                        1   
..                  ...              ...                      ...   
694              776715                3                        1   
695              841769                2                        1   
696              888820                5                       10   
697              897471                4                        8   
698              897471                4                        8   

     Uniformity of cell shape  marginal Ahesion  Single epithelial cell size  \
0                           1                 1                            2   
1                           4                 5                            7   
2                           1                 1                            2   
3                           8                 1                            3   
4                           1                 3                            2   
..                        ...               ...                          ...   
694                         1                 1                            3   
695                         1                 1                            2   
696                        10                 3                            7   
697                         6                 4                            3   
698                         8                 5                            4   

    Bare Nuclei  Bland Chromatin  Normal Nucleoli  Mitoses  Class  
0             1                3                1        1      2  
1            10                3                2        1      2  
2             2                3                1        1      2  
3             4                3                7        1      2  
4             1                3                1        1      2  
..          ...              ...              ...      ...    ...  
694           2                1                1        1      2  
695           1                1                1        1      2  
696           3                8               10        2      4  
697           4               10                6        1      4  
698           5               10                4        1      4  

[699 rows x 11 columns]

In [4]:
bc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 699 entries, 0 to 698
Data columns (total 11 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   Sample code number           699 non-null    int64 
 1   Clump thickness              699 non-null    int64 
 2   Uniformity of cell Size      699 non-null    int64 
 3   Uniformity of cell shape     699 non-null    int64 
 4   marginal Ahesion             699 non-null    int64 
 5   Single epithelial cell size  699 non-null    int64 
 6   Bare Nuclei                  699 non-null    object
 7   Bland Chromatin              699 non-null    int64 
 8   Normal Nucleoli              699 non-null    int64 
 9   Mitoses                      699 non-null    int64 
 10  Class                        699 non-null    int64 
dtypes: int64(10), object(1)
memory usage: 60.2+ KB


In [5]:
bc.describe()

Sample code number  Clump thickness  Uniformity of cell Size  \
count        6.990000e+02       699.000000               699.000000   
mean         1.071704e+06         4.417740                 3.134478   
std          6.170957e+05         2.815741                 3.051459   
min          6.163400e+04         1.000000                 1.000000   
25%          8.706885e+05         2.000000                 1.000000   
50%          1.171710e+06         4.000000                 1.000000   
75%          1.238298e+06         6.000000                 5.000000   
max          1.345435e+07        10.000000                10.000000   

       Uniformity of cell shape  marginal Ahesion  \
count                699.000000        699.000000   
mean                   3.207439          2.806867   
std                    2.971913          2.855379   
min                    1.000000          1.000000   
25%                    1.000000          1.000000   
50%                    1.000000          1.000000   
75%                    5.000000          4.000000   
max                   10.000000         10.000000   

       Single epithelial cell size  Bland Chromatin  Normal Nucleoli  \
count                   699.000000       699.000000       699.000000   
mean                      3.216023         3.437768         2.866953   
std                       2.214300         2.438364         3.053634   
min                       1.000000         1.000000         1.000000   
25%                       2.000000         2.000000         1.000000   
50%                       2.000000         3.000000         1.000000   
75%                       4.000000         5.000000         4.000000   
max                      10.000000        10.000000        10.000000   

          Mitoses       Class  
count  699.000000  699.000000  
mean     1.589413    2.689557  
std      1.715078    0.951273  
min      1.000000    2.000000  
25%      1.000000    2.000000  
50%      1.000000    2.000000  
75%      1.000000    4.000000  
max     10.000000    4.000000

In [6]:
bc['Bare Nuclei'].value_counts()['?']

16

In [7]:
bc['Bare Nuclei'] = pd.to_numeric(bc['Bare Nuclei'], errors='coerce')

In [8]:
bc['Bare Nuclei'].fillna(bc['Bare Nuclei'].mean(), inplace = True)

In [9]:
for i in range (0,11):
    bc[col_Names[i]] = pd.to_numeric(bc[col_Names[i]], errors='coerce')
    print(bc[col_Names[i]].isna().sum())

0
0
0
0
0
0
0
0
0
0
0


In [10]:
bc.corr()

Sample code number  Clump thickness  \
Sample code number                     1.000000        -0.055308   
Clump thickness                       -0.055308         1.000000   
Uniformity of cell Size               -0.041603         0.644913   
Uniformity of cell shape              -0.041576         0.654589   
marginal Ahesion                      -0.064878         0.486356   
Single epithelial cell size           -0.045528         0.521816   
Bare Nuclei                           -0.098668         0.587300   
Bland Chromatin                       -0.060051         0.558428   
Normal Nucleoli                       -0.052072         0.535835   
Mitoses                               -0.034901         0.350034   
Class                                 -0.080226         0.716001   

                             Uniformity of cell Size  \
Sample code number                         -0.041603   
Clump thickness                             0.644913   
Uniformity of cell Size                     1.000000   
Uniformity of cell shape                    0.906882   
marginal Ahesion                            0.705582   
Single epithelial cell size                 0.751799   
Bare Nuclei                                 0.686801   
Bland Chromatin                             0.755721   
Normal Nucleoli                             0.722865   
Mitoses                                     0.458693   
Class                                       0.817904   

                             Uniformity of cell shape  marginal Ahesion  \
Sample code number                          -0.041576         -0.064878   
Clump thickness                              0.654589          0.486356   
Uniformity of cell Size                      0.906882          0.705582   
Uniformity of cell shape                     1.000000          0.683079   
marginal Ahesion                             0.683079          1.000000   
Single epithelial cell size                  0.719668          0.599599   
Bare Nuclei                                  0.709606          0.665049   
Bland Chromatin                              0.735948          0.666715   
Normal Nucleoli                              0.719446          0.603352   
Mitoses                                      0.438911          0.417633   
Class                                        0.818934          0.696800   

                             Single epithelial cell size  Bare Nuclei  \
Sample code number                             -0.045528    -0.098668   
Clump thickness                                 0.521816     0.587300   
Uniformity of cell Size                         0.751799     0.686801   
Uniformity of cell shape                        0.719668     0.709606   
marginal Ahesion                                0.599599     0.665049   
Single epithelial cell size                     1.000000     0.581261   
Bare Nuclei                                     0.581261     1.000000   
Bland Chromatin                                 0.616102     0.675896   
Normal Nucleoli                                 0.628881     0.577362   
Mitoses                                         0.479101     0.338740   
Class                                           0.682785     0.816050   

                             Bland Chromatin  Normal Nucleoli   Mitoses  \
Sample code number                 -0.060051        -0.052072 -0.034901   
Clump thickness                     0.558428         0.535835  0.350034   
Uniformity of cell Size             0.755721         0.722865  0.458693   
Uniformity of cell shape            0.735948         0.719446  0.438911   
marginal Ahesion                    0.666715         0.603352  0.417633   
Single epithelial cell size         0.616102         0.628881  0.479101   
Bare Nuclei                         0.675896         0.577362  0.338740   
Bland Chromatin                     1.000000         0.665878  0.344169   
Normal Nucleoli                     0.665878         1.000000  0.428336   
Mitoses    

##### As we can see, that the correlation between target variable and sample code number is negative. So in the next step we are dropping sample code number from the dataframe.

In [11]:
bc.drop("Sample code number",axis=1,inplace=True) 

In [12]:
col_Names=["Clump thickness", "Uniformity of cell Size", "Uniformity of cell shape", "marginal Ahesion","Single epithelial cell size", "Bare Nuclei", "Bland Chromatin", "Normal Nucleoli", "Mitoses"]


In [13]:
for i in col_Names:
    
    bc[(bc[col_Names] > bc[col_Names].mean() + 3*bc[col_Names].std()) | (bc[col_Names] < bc[col_Names].mean() - 3*bc[col_Names].std())]

In [14]:
for i in col_Names:
    
    upper_limit = bc[col_Names].mean() + 3*bc[col_Names].std()
    lower_limit = bc[col_Names].mean() - 3*bc[col_Names].std()
    bc[col_Names] = np.where(
        bc[col_Names]>upper_limit,
        upper_limit,
        np.where(
            bc[col_Names]<lower_limit,
            lower_limit,
            bc[col_Names]
        )
    )

In [15]:
bc.shape


(699, 10)

In [16]:
bc.dtypes

Clump thickness                float64
Uniformity of cell Size        float64
Uniformity of cell shape       float64
marginal Ahesion               float64
Single epithelial cell size    float64
Bare Nuclei                    float64
Bland Chromatin                float64
Normal Nucleoli                float64
Mitoses                        float64
Class                            int64
dtype: object

In [17]:
X = bc.drop(columns= 'Class')
y = bc.Class

In [18]:
X

Clump thickness  Uniformity of cell Size  Uniformity of cell shape  \
0                5.0                      1.0                       1.0   
1                5.0                      4.0                       4.0   
2                3.0                      1.0                       1.0   
3                6.0                      8.0                       8.0   
4                4.0                      1.0                       1.0   
..               ...                      ...                       ...   
694              3.0                      1.0                       1.0   
695              2.0                      1.0                       1.0   
696              5.0                     10.0                      10.0   
697              4.0                      8.0                       6.0   
698              4.0                      8.0                       8.0   

     marginal Ahesion  Single epithelial cell size  Bare Nuclei  \
0                 1.0                          2.0          1.0   
1                 5.0                          7.0         10.0   
2                 1.0                          2.0          2.0   
3                 1.0                          3.0          4.0   
4                 3.0                          2.0          1.0   
..                ...                          ...          ...   
694               1.0                          3.0          2.0   
695               1.0                          2.0          1.0   
696               3.0                          7.0          3.0   
697               4.0                          3.0          4.0   
698               5.0                          4.0          5.0   

     Bland Chromatin  Normal Nucleoli  Mitoses  
0                3.0              1.0      1.0  
1                3.0              2.0      1.0  
2                3.0              1.0      1.0  
3                3.0              7.0      1.0  
4                3.0              1.0      1.0  
..               ...              ...      ...  
694              1.0              1.0      1.0  
695              1.0              1.0      1.0  
696              8.0             10.0      2.0  
697             10.0              6.0      1.0  
698             10.0              4.0      1.0  

[699 rows x 9 columns]

In [19]:
y

0      2
1      2
2      2
3      2
4      2
      ..
694    2
695    2
696    4
697    4
698    4
Name: Class, Length: 699, dtype: int64

In [20]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size = 0.25, stratify = y, random_state = 1)
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((524, 9), (175, 9), (524,), (175,))

## Decision Tree

In [21]:
from sklearn.tree import DecisionTreeClassifier

dt = DecisionTreeClassifier(random_state=1)

dt.fit(X_train, y_train)

pred = dt.predict(X_test)

dt_score = dt.score(X_test, y_test)

result = pd.DataFrame({'Model' : ['Decision Tree Classifier'], 'Accuracy' : [dt_score]})

result

Model  Accuracy
0  Decision Tree Classifier  0.948571

## Decision Tree with K fold 

In [22]:
from sklearn.model_selection import cross_val_score
scores= cross_val_score(dt,X,y,cv=10,scoring='accuracy')

In [23]:
scores

array([0.88571429, 0.92857143, 0.94285714, 0.9       , 0.97142857,
       0.95714286, 0.95714286, 0.98571429, 0.98571429, 0.98550725])

In [24]:
scores.mean()

0.9499792960662529

#### We observe that there is 0.001 increase in accuracy when used k-fold cross validation

## Random Forest

In [25]:
from sklearn.ensemble import RandomForestClassifier


In [26]:
rf = RandomForestClassifier(n_estimators=10,random_state=1)
scores_rf= cross_val_score(rf,X,y,cv=10,scoring='accuracy')

In [27]:
scores_rf

array([0.87142857, 0.91428571, 0.95714286, 0.9       , 1.        ,
       0.95714286, 0.91428571, 0.98571429, 0.97142857, 0.98550725])

In [28]:
scores_rf.mean()
result.loc[1] = ['Random Forest', scores_rf.mean()]

In [29]:
result

Model  Accuracy
0  Decision Tree Classifier  0.948571
1             Random Forest  0.945694

### Does it perform better than decision tree?

##### Yes Decision Tree performed better than Random Forest by a difference of 0.003 in acccuracy

## KNN 

In [44]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import cross_val_score


TypeError: Singleton array array(Ellipsis, dtype=object) cannot be considered a valid collection.

In [43]:
 knn = KNeighborsClassifier(n_neighbors= 10)

In [32]:
scores_knn= cross_val_score(knn,X,y,cv=10,scoring='accuracy')

C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `kee

In [33]:
scores_knn

array([0.9       , 0.95714286, 0.95714286, 0.91428571, 1.        ,
       0.97142857, 0.97142857, 0.98571429, 0.98571429, 1.        ])

In [34]:
scores_knn.mean()

0.9642857142857142

In [35]:
x=0
k=0
for i in range (1,51):
    knn=KNeighborsClassifier(n_neighbors= i)
    
    scores_knn= cross_val_score(knn,X,y,cv=10,scoring='accuracy')
    
    scores_knn.mean()
    
    print("For k = "+ str(i) +"- Acccuracy = "+str(scores_knn.mean()))
    
    if(scores_knn.mean()>x):
        x=scores_knn.mean()
        k=i

print("The best accuracy levels are achieved when k is "+ str(k))

C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `kee

C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `kee

For k = 1- Acccuracy = 0.9499585921325051
For k = 2- Acccuracy = 0.9413871635610767
For k = 3- Acccuracy = 0.9642443064182193
For k = 4- Acccuracy = 0.9599585921325051
For k = 5- Acccuracy = 0.9685507246376812


C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `kee

C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `kee

For k = 6- Acccuracy = 0.9642650103519669
For k = 7- Acccuracy = 0.9685714285714285
For k = 8- Acccuracy = 0.9642857142857142
For k = 9- Acccuracy = 0.9671428571428571


C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `kee

C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `kee

C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `kee

For k = 10- Acccuracy = 0.9642857142857142
For k = 11- Acccuracy = 0.9657142857142859
For k = 12- Acccuracy = 0.962857142857143
For k = 13- Acccuracy = 0.9657142857142859
For k = 14- Acccuracy = 0.9614285714285714


C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `kee

C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `kee

For k = 15- Acccuracy = 0.9642857142857142
For k = 16- Acccuracy = 0.962857142857143
For k = 17- Acccuracy = 0.962857142857143
For k = 18- Acccuracy = 0.962857142857143
For k = 19- Acccuracy = 0.962857142857143


C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `kee

C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `kee

For k = 20- Acccuracy = 0.9614285714285714
For k = 21- Acccuracy = 0.9614285714285714
For k = 22- Acccuracy = 0.9600000000000002
For k = 23- Acccuracy = 0.9600000000000002
For k = 24- Acccuracy = 0.9585714285714285


C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `kee

C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `kee

For k = 25- Acccuracy = 0.9585714285714285
For k = 26- Acccuracy = 0.9585714285714285
For k = 27- Acccuracy = 0.9585714285714285
For k = 28- Acccuracy = 0.9600000000000002
For k = 29- Acccuracy = 0.9600000000000002


C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `kee

For k = 30- Acccuracy = 0.9600000000000002
For k = 31- Acccuracy = 0.9600000000000002
For k = 32- Acccuracy = 0.9585714285714285
For k = 33- Acccuracy = 0.9585714285714285


C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `kee

C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `kee

For k = 34- Acccuracy = 0.9585714285714285
For k = 35- Acccuracy = 0.9585714285714285
For k = 36- Acccuracy = 0.96
For k = 37- Acccuracy = 0.96
For k = 38- Acccuracy = 0.9585714285714285


C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `kee

For k = 39- Acccuracy = 0.9585714285714285
For k = 40- Acccuracy = 0.9585714285714285
For k = 41- Acccuracy = 0.9585714285714285
For k = 42- Acccuracy = 0.9585714285714285
For k = 43- Acccuracy = 0.9585714285714285


C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `kee

C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `kee

For k = 44- Acccuracy = 0.9585714285714285
For k = 45- Acccuracy = 0.9585714285714285
For k = 46- Acccuracy = 0.9571428571428571
For k = 47- Acccuracy = 0.9585714285714285
For k = 48- Acccuracy = 0.9571428571428571
For k = 49- Acccuracy = 0.9585714285714285
For k = 50- Acccuracy = 0.9585714285714285
The best accuracy levels are achieved when k is 7


C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
C:\Users\Laghavi\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `kee

## Naives Bayes

In [36]:
from sklearn.naive_bayes import MultinomialNB

mnb = MultinomialNB()

scores_mnb= cross_val_score(mnb,X,y,cv=10,scoring='accuracy')

scores_mnb.mean()


0.9013871635610766

In [37]:
from sklearn.naive_bayes import GaussianNB
gnb = GaussianNB()

scores_gnb= cross_val_score(gnb,X,y,cv=10,scoring='accuracy')

scores_gnb.mean()

0.9599585921325053

### Does Gaussian perform better than MultinomialNB?

##### Yes, Gaussian performed better than MultinominalNB with difference of 0.05 in accuracy

In [38]:
result.loc[2] = ['KNN', scores_knn.mean()]
result.loc[3] = ['MultinominalNB', scores_mnb.mean()]
result.loc[4] = ['GaussianNB', scores_gnb.mean()]

result

Model  Accuracy
0  Decision Tree Classifier  0.948571
1             Random Forest  0.945694
2                       KNN  0.958571
3            MultinominalNB  0.901387
4                GaussianNB  0.959959

## SVM 

In [39]:
from sklearn.svm import SVC


In [40]:
j=5
svm_kernels=["linear","sigmoid","poly"]
for i in svm_kernels:
    svm_classifier = SVC(kernel=i,random_state=1)
    
    cv_scores_1 = cross_val_score(svm_classifier,X,y,scoring='accuracy',cv=10)
    print("When kernel = "+i)
    print(cv_scores_1)
    print(np.mean(cv_scores_1))
    
    result.loc[j] = ['svm.SVC when kernel = '+i, np.mean(cv_scores_1)]
    j+=1

When kernel = linear
[0.91428571 0.97142857 0.97142857 0.91428571 0.98571429 0.95714286
 0.95714286 1.         0.97142857 1.        ]
0.9642857142857142
When kernel = sigmoid
[0.35714286 0.3        0.48571429 0.48571429 0.4        0.3
 0.48571429 0.48571429 0.45714286 0.44927536]
0.4206418219461698
When kernel = poly
[0.85714286 0.95714286 0.94285714 0.91428571 1.         0.95714286
 0.98571429 1.         0.98571429 0.94202899]
0.9542028985507247


In [41]:
result

Model  Accuracy
0       Decision Tree Classifier  0.948571
1                  Random Forest  0.945694
2                            KNN  0.958571
3                 MultinominalNB  0.901387
4                     GaussianNB  0.959959
5   svm.SVC when kernel = linear  0.964286
6  svm.SVC when kernel = sigmoid  0.420642
7     svm.SVC when kernel = poly  0.954203

##### When the Kernel is Linear, we got the highest accuracy of 96.4%

# Conclusion

#### Out of all the models, we got highest accuracy for svm when the kernel is linear. 